File: object-oriented-programming/virtual-world-python/animals/wolf.py
from animals.animal import Animal
from species import Species
import random

class Wolf(Animal) :

    def __init__(self, world, x = None, y = None) :
        if x == None and y == None :
            super().__init__(Species.WOLF, 0, 0, 9, 5, world)
            width = world.width
            height = world.height

            while(True):
                new_x = random.randint(0, width - 1)
                new_y = random.randint(0, height - 1)
                if self._world.is_free(new_x, new_y) :
                    break
                else :
                    continue

            self._world.set_organism(None, self.x, self.y)
            self._x = new_x
            self._y = new_y
            self._world.set_organism(self, self.x, self.y)
            return
        
        super().__init__(Species.WOLF, x, y, 9, 5, world)

    def action(self) :
        return super().action()

    def collision(self, other):
        return super().collision(other)

    def draw_color(self):
        return self.species.color
    
    def breed(self, x, y):
        offspring = Wolf(self._world, x, y)
        offspring.set_breed_cooldown()
        return offspring

File: object-oriented-programming/virtual-world-cpp/src/animals/Wolf.cpp
#include "Wolf.h"
#include "World.h"
#include "Species.h"
#include <iostream>

Wolf::Wolf(int x, int y, World* world) 
    : Animal(Species::WOLF, x, y, 9, 5, world, 0, true) {}

Wolf::Wolf(World* world)
    : Animal(Species::WOLF, 0, 0, 9, 5, world, 0, true) 
    {
        int rows = world->getRows();
        int columns = world->getColumns();

        int newX, newY;

        do{
            newX = rand() % rows;
            newY = rand() % columns;
        } while(!world->isFree(newX, newY));

        world->setOrganism(nullptr, x, y);
        x = newX;
        y = newY;
        world->setOrganism(this, x, y);
    }

Wolf::~Wolf() {}

char Wolf::draw() const {
    return 'W';
}

Wolf* Wolf::breed(int x, int y) const {
    Wolf* offspring = new Wolf(x, y, world);
    offspring->setBreedCooldown();
    return offspring;
}

Wolf* Wolf::loadFrom(World* world, std::istream &in) {
    int s_x, s_y, s_strength, s_initiative, s_age, s_breedCooldown, s_canBreed;
    in >> s_x >> s_y >> s_strength >> s_initiative >> s_age >> s_breedCooldown >> s_canBreed;
    Wolf* wolf = new Wolf(s_x, s_y, world);
    wolf->strength = s_strength;
    wolf->initiative = s_initiative;
    wolf->age = s_age;
    wolf->breedCooldown = s_breedCooldown;
    wolf->canBreed = bool(s_canBreed);
    return wolf;
}